src/com/amazon/ionelement/api/StructUpdateBuilder.kt

```diff
@@ -0,0 +1,121 @@
+package com.amazon.ionelement.api
+
+import com.amazon.ionelement.api.StructUpdateBuilder.FieldUpdateMode.*
+import com.amazon.ionelement.impl.*
+import kotlinx.collections.immutable.toPersistentList
+import kotlinx.collections.immutable.toPersistentMap
+
+/**
+ * An [IonPath] represents a traversal into a nested IonElement made of [StructElement]s and [SeqElement]s
+ */
+public class IonPath(public val elements: List<IonPathElement>) : Collection<IonPathElement> by elements {
+    public constructor(vararg elements: IonPathElement) : this(elements.asList())
+
+    public fun tail(): IonPath {
+        return IonPath(elements.drop(1))
+    }
+}
+
+public sealed class IonPathElement
+
+/**
+ * Traverse into a nested [StructElement] by its field name.
+ */
+public data class Field(val fieldName: String) : IonPathElement()
+
+/**
+ * Traverse into a nested [SeqElement] by its numerical index.
+ */
+public data class Index(val index: Int) : IonPathElement()
+
+/**
+ * Fetch an element from a nested IonElement using the [path] to traverse struct fields and [SeqElement] indices
```

Comment:
The `getPath` was left over from some other experiments I was doing and I think it makes sense to move to another PR. 

> What happens if there is more than one element matching the path?

There should be separate functions for getting the first value and getting all values, similar to the existing API for StructElement.

> Would you consider adding a getPathOrNull function—similar to how the Kotlin Standard Library first() and firstOrNull()?

Yes, makes sense.

> What would you think of adding a type parameter to the function so that you can get a specific type of element from the given path?

Great idea, this seems idiomatic. If the result does not match up I suppose an `IonElementConstraintException` should be thrown similar to the `as*` functions.

I had some alternative approaches for representing the path as well. I will make a separate PR and move the discussion there. I'll keep this PR for struct updating.